### PR TITLE
[perf, parallel] feat: add CUDA graph capture for text decoder modules

### PR DESCRIPTION
**Pull request by wang-zerui**
### What does this PR do?

Adds CUDA graph capture support for text decoder modules.

This implements the training-side CUDA graph path discussed in RFC #401:
https://github.com/ByteDance-Seed/VeOmni/issues/401

The change adds `train.cuda_graph.*` arguments, applies lazy shape-specialized CUDA graph capture after model parallelization, and supports distributed training with FSDP2 by replaying graphs with the current module parameters and buffers as explicit graph inputs.

> This PR replaces the torch.compile-based exploration in [#810](https://github.com/ByteDance-Seed/VeOmni/pull/810) with explicit CUDA graph capture, which, However, has unresolved precision issue and leads to training loss mismatch.

### Checklist Before Starting

- Search for relative PRs/issues and link here: https://github.com/ByteDance-Seed/VeOmni/issues/401
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

A/B training validation, 500 steps, Qwen2.5-style 72B config, 8x GB200 GPUs across 2 nodes, FSDP2 data parallel size 8, seq_len=128, global_batch_size=8:

| Setting | Avg train step time | Tokens/s | Max loss diff |
| --- | ---: | ---: | ---: |
| CUDA graph off | 2.96s | 345.95 | - |
| CUDA graph on | 2.69s | 380.67 | 0.0 |

Observed speedup: **1.10x**.

Repro command pattern:

```bash
# Baseline
NPROC_PER_NODE=4 bash train.sh tasks/train_text.py configs/text/qwen2_5.yaml \
  --data.max_seq_len 128 \
  --data.train_size 4096000 \
  --train.max_steps 500 \
  --train.global_batch_size 8 \
  --train.micro_batch_size 1 \
  --train.gradient_checkpointing.enable false \
  --train.pad_to_length true \
  --train.cuda_graph.enable false \
  --train.checkpoint.save_steps 0 \
  --train.checkpoint.hf_save_steps 0

# CUDA graph
NPROC_PER_NODE=4 bash train.sh tasks/train_text.py configs/text/qwen2_5.yaml \
  --data.max_seq_len 128 \
  --data.train_size 4096000 \
  --train.max_steps 500 \
  --train.global_batch_size 8 \
  --train.micro_batch_size 1 \
  --train.gradient_checkpointing.enable false \
  --train.pad_to_length true \
  --train.cuda_graph.enable true \
  --train.cuda_graph.scope auto \
  --train.checkpoint.save_steps 0 \
  --train.checkpoint.hf_save_steps 0
```

Logged scalar loss matched exactly across all 500 steps. The saved logs are identical for all emitted decimal digits, with max/mean absolute logged loss diff = 0.0.

image

Pre-commit checks passed:

```text
check python ast                    Passed
check for added large files          Passed
check for merge conflicts            Passed
check yaml                           Skipped
debug statements                     Passed
fix end of files                     Passed
fix requirements.txt                 Skipped
trim trailing whitespace             Passed
don't commit to branch               Passed
pyupgrade                            Passed
ruff                                 Passed
ruff format                          Passed
```

### API and Usage Example

New training arguments:

```yaml
train:
  cuda_graph:
    enable: true
    scope: auto
    num_warmup_steps: 3
    max_graphs_per_layer: 16
    strict: false
```

`scope` options:

- `auto`: capture dense decoder layers; capture attention modules for MoE decoder layers with dynamic expert routing.
- `layer`: capture whole decoder layers. For MoE decoder layers, this requires static expert inputs.
- `attn`: capture attention modules only.

CUDA graph capture is supported for text-only models and cannot be combined with gradient checkpointing.

### Design & Code Changes

- Adds `CudaGraphConfig` and argument validation for CUDA graph capture.
- Adds lazy, shape-specialized decoder module capture in `veomni.distributed.cuda_graph`.
- Applies CUDA graph wrapping after FSDP/DDP/TP parallelization.
- Uses explicit parameter and buffer graph inputs for FSDP2 replay.
- Validates that CUDA graph capture is enabled only for text-only model configs.
- Adds `router_aux_loss_mask` collation for Qwen3.5-MoE so router auxiliary loss ignores padded tokens.
- Updates generated Qwen3.5-MoE patch files to consume `router_aux_loss_mask` when present.
- Documents `train.cuda_graph.*` arguments and padding interaction.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

File: veomni/models/transformers/qwen2/qwen2_gpu_patch_gen_config.py [hunk 1]
```diff
@@ -168,10 +179,20 @@ def qwen2_model_forward_patched(
             "position_ids": position_ids,
         }
         causal_mask_mapping = {
-            "full_attention": create_causal_mask(**mask_kwargs),
+            "full_attention": _veomni_create_mask(
```
Comment (FoolPlayer) [inline veomni/models/transformers/qwen2/qwen2_gpu_patch_gen_config.py:182]: Why import `_veomni_create_mask`, it seems unrelated with cuda graph.

File: veomni/arguments/__init__.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 from .arguments_types import (
     AcceleratorConfig,
     CheckpointConfig,
+    CudaGraphConfig,
```
Comment (FoolPlayer) [inline veomni/arguments/__init__.py:19]: Maybe we can use `CompileConfig`, cause we may have different backend like Ascend

File: veomni/distributed/cuda_graph.py [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+    "video_token_id",
+    "audio_token_id",
+    "vision_token_id",
+    "vision_start_token_id",
+    "vision_end_token_id",
+    "image_token_index",
+    "video_token_index",
+    "audio_token_index",
+)
+_NON_TEXT_NAME_MARKERS = (
```
Comment (FoolPlayer) [inline veomni/distributed/cuda_graph.py:63]: Maybe these markers are too weak?

File: veomni/distributed/cuda_graph.py [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+            self._warn_once(
+                state,
+                f"Disable CUDA graph for {self.module.__class__.__name__}: {exc}.",
+            )
+            return self._fallback(*args, **kwargs)
+
+        state = self.graphs.get(key)
+        if state is None:
+            if len(self.graphs) >= self.max_graphs:
+                return self._fallback(*args, **kwargs)
```
Comment (FoolPlayer) [inline veomni/distributed/cuda_graph.py:519]: Add warning.

**Comment by FoolPlayer** [comment]
Thanks for you excellent work! But need some fix.
1. Please add fsdp e2e ci with cuda-graph for both moe/dense model with sp/ep.
2. Can you add MoE training task example with compile.
3. For the results you attached, it seems strange, the diff for loss is totally 0. I think byte-identical is not that possible, it seems that the logged loss was truncated to ~6 significant digits. Can you have a check?
